bundler/spec/install/gemfile/sources_spec.rb

```diff
@@ -371,7 +371,15 @@
 
       it "fails" do
         bundle :install, :artifice => "compact_index", :raise_on_error => false
-        expect(err).to include("Could not find gem 'missing', which is required by gem 'depends_on_missing', in any of the sources.")
+        expect(err).to end_with <<~E.strip
+          Could not find compatible versions
+
+          Because every version of depends_on_missing depends on missing >= 0
+            and no versions in any of the sources satisfy missing >= 0,
+            every version of depends_on_missing is forbidden.
+          So, because Gemfile depends on depends_on_missing >= 0,
+            version solving has failed.
+        E
```

Comment:
I'm not sure if this kind of message is clearer or not. "No versions satisfy `gem` >= 0" seems like an unclear way of saying "we could not find a gem named `gem` in any sources or installed locally", to me.

Reply:
I think there may be room for improvement here.

 This is original pub grub message:

```
Because every version of depends_on_missing depends on missing >= 0
  and no versions satisfy missing >= 0,
  every version of depends_on_missing is forbidden.
So, because Gemfile depends on depends_on_missing >= 0,
  version solving has failed.
```

I added a small tweak to mention the specific source for the gem when no versions could be found, hence the **in any of the sources** extra bit. Note that this case is particularly unspecific ("in any of the sources")  because it's a spec for the legacy multiple global sources mode. Normally, error messages will include a specific source ("in `https://rubygems.org` or installed locally").

I figured mentioning the source could sometimes help figuring out why no matching versions were found, since that's we already do today.

So, your suggestion would be something like the following?

```
Because every version of depends_on_missing depends on missing >= 0
  and missing >= 0 could not be found in any of the sources,
  every version of depends_on_missing is forbidden.
So, because Gemfile depends on depends_on_missing >= 0,
  version solving has failed.
```

Reply:
Yes, exactly, something like that. I might say "cannot be used" instead of "forbidden"? But that's basically what I was thinking. :)

Reply:
Great! I pushed updated messages.

The "forbidden" part is a bit trickier because pub grub we'd need to make sure it reads well in every situation, and it may get a bit complex. See https://github.com/jhawthorn/pub_grub/blob/9d8e9a5d2ce781e3967cffe1cc9ca3838613b457/lib/pub_grub/incompatibility.rb#L55-L102.

Reply:
@indirect Do you think it's worth re-ticketing the suggestion to avoid "forbidden" in favor of other wording? I guess we'd need to change two cases:

#### With a universal constraint:

Change

> every version of depends_on_missing is forbidden

to 

> depends_on_missing cannot be used

#### With another constraint:

Change

> depends_on_missing >= 2.0 is forbidden

to

> depends_on_missing >= 2.0 cannot be used

I'm happy to open an issue at pub_grub upstream.

The proposed missing feels better to me but I don't know why, so hard to justify changing. Maybe "forbidden" is too strong of a word?

Reply:
A separate ticket sounds fine to me. I think the reason “forbidden” can possibly be a little misleading is because we often use that word around permissions or authorization—so it might give the impressions that the user could use that gem if only they had the right permission. Instead, we are trying to say “we can’t do that because it is a logical impossibility” which has a different feel.

Reply:
Understood! Created an issue at https://github.com/jhawthorn/pub_grub/issues/18.